Userland/Libraries/LibGUI/TextDocument.cpp

```diff
@@ -383,6 +407,96 @@ DeprecatedString TextDocument::text_in_range(TextRange const& a_range) const
 }
 
+// This function will determine the number of codepoints of the glyph immediately preceding the cursor, by determining if there is an emoji. If no emoji is found it will return 1.
+size_t TextDocument::get_code_points_before_cursor(TextPosition const& cursor) const
+{
+    if (!cursor.is_valid())
+        return 0;
+
+    return get_prev_grapheme_cluster_boundary(cursor);
+}
+
+// This function will return the position of the next grapheme cluster break, relative to the cursor, for "correct looking" parsing of unicode based on grapheme cluster boundary algorithm.
+size_t TextDocument::get_next_grapheme_cluster_boundary(TextPosition const& cursor) const
+{
+    if (!cursor.is_valid())
+        return 0;
+
+    auto& line = this->line(cursor.line());
+    auto line_span = Span<u32 const>(line.code_points(), line.length());
+
+    if (line_span.is_empty())
+        return 0;
+
+    if (cursor.column() == line.length())
+        return 0;
+
+    auto boundaries_cache = Unicode::find_grapheme_segmentation_boundaries(Utf32View(line.code_points(), line.length()));
+
+    if (boundaries_cache.is_empty())
+        return cursor.column();
+
+    if (boundaries_cache.size() == 1)
+        return cursor.column() - boundaries_cache.at(0);
+
+    auto previous_boundary_guess = boundaries_cache.at(0);
+    auto boundary_guess = boundaries_cache.at(1);
+
+    for (auto current_boundary_guess : boundaries_cache) {
+        // If the current boundary is further to the left, update our guess
+        if (current_boundary_guess < boundary_guess) {
+            previous_boundary_guess = boundary_guess;
+            boundary_guess = current_boundary_guess;
+        }
+    }
+
+    // Assume that the cursor cannot sit at a point that is not a boundary, so return previous guess.
+    return previous_boundary_guess;
+}
+
+// This function will return the position of the previous grapheme cluster break, relative to the cursor, for "correct looking" parsing of unicode based on grapheme cluster boundary algorithm.
+size_t TextDocument::get_prev_grapheme_cluster_boundary(TextPosition const& cursor) const
+{
+    if (!cursor.is_valid())
+        return 0;
+
+    auto& line = this->line(cursor.line());
+    auto line_span = Span<u32 const>(line.code_points(), line.length());
+
+    if (line_span.is_empty())
+        return 0;
+
+    if (cursor.column() < 2)
+        return cursor.column();
+
+    auto boundaries_cache = Unicode::find_grapheme_segmentation_boundaries(Utf32View(line.code_points(), line.length()));
+    if (boundaries_cache.is_empty())
+        return cursor.column();
+
+    if (boundaries_cache.size() == 1)
+        return cursor.column() - boundaries_cache.at(0);
+
+    auto boundary_guess = boundaries_cache.at(0);
+
+    for (auto current_boundary_guess : boundaries_cache) {
+        // If the current boundary is further to the right, update our guess.
+        if ((current_boundary_guess - cursor.column()) > (boundary_guess - cursor.column()))
```

Comment:
This is a bit confusing to read, can we write this without incurring a bunch of overflow? Would something like this suffice?
```c++
    for (auto current_boundary_guess : boundaries_cache) {
        if (current_boundary_guess < cursor.column())
            boundary_guess = current_boundary_guess;
        else
            break;
    }
```